File: apps/api/internal/handler/systemsettings.go
package handler

import (
	"encoding/json"
	"errors"
	"net/http"
	"time"

	"github.com/go-openapi/strfmt"
	"github.com/google/uuid"

	"github.com/tolga/terp/gen/models"
	"github.com/tolga/terp/internal/middleware"
	"github.com/tolga/terp/internal/model"
	"github.com/tolga/terp/internal/service"
)

// SystemSettingsHandler handles system settings HTTP requests.
type SystemSettingsHandler struct {
	svc          *service.SystemSettingsService
	auditService *service.AuditLogService
}

// NewSystemSettingsHandler creates a new SystemSettingsHandler.
func NewSystemSettingsHandler(svc *service.SystemSettingsService) *SystemSettingsHandler {
	return &SystemSettingsHandler{svc: svc}
}

// SetAuditService sets the audit log service for this handler.
func (h *SystemSettingsHandler) SetAuditService(s *service.AuditLogService) {
	h.auditService = s
}

// GetSettings handles GET /system-settings
func (h *SystemSettingsHandler) GetSettings(w http.ResponseWriter, r *http.Request) {
	tenantID, ok := middleware.TenantFromContext(r.Context())
	if !ok {
		respondError(w, http.StatusUnauthorized, "Tenant required")
		return
	}

	settings, err := h.svc.Get(r.Context(), tenantID)
	if err != nil {
		respondError(w, http.StatusInternalServerError, "Failed to get system settings")
		return
	}

	respondJSON(w, http.StatusOK, mapSystemSettingsToResponse(settings))
}

// UpdateSettings handles PUT /system-settings
func (h *SystemSettingsHandler) UpdateSettings(w http.ResponseWriter, r *http.Request) {
	tenantID, ok := middleware.TenantFromContext(r.Context())
	if !ok {
		respondError(w, http.StatusUnauthorized, "Tenant required")
		return
	}

	var req models.UpdateSystemSettingsRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		respondError(w, http.StatusBadRequest, "Invalid request body")
		return
	}

	input := service.UpdateSystemSettingsInput{}

	// Map boolean fields (use pointer nil to detect presence)
	if req.RoundingRelativeToPlan {
		v := req.RoundingRelativeToPlan
		input.RoundingRelativeToPlan = &v
	}
	if req.ErrorListEnabled {
		v := req.ErrorListEnabled
		input.ErrorListEnabled = &v
	}
	if req.AutoFillOrderEndBookings {
		v := req.AutoFillOrderEndBookings
		input.AutoFillOrderEndBookings = &v
	}
	if req.FollowUpEntriesEnabled {
		v := req.FollowUpEntriesEnabled
		input.FollowUpEntriesEnabled = &v
	}
	if req.ProxyEnabled {
		v := req.ProxyEnabled
		input.ProxyEnabled = &v
	}
	if req.ServerAliveEnabled {
		v := req.ServerAliveEnabled
		input.ServerAliveEnabled = &v
	}
	if req.ServerAliveNotifyAdmins {
		v := req.ServerAliveNotifyAdmins
		input.ServerAliveNotifyAdmins = &v
	}
	if req.TrackedErrorCodes != nil {
		input.TrackedErrorCodes = req.TrackedErrorCodes
	}

	// Map optional integer fields
	if req.BirthdayWindowDaysBefore != nil {
		v := int(*req.BirthdayWindowDaysBefore)
		input.BirthdayWindowDaysBefore = &v
	}
	if req.BirthdayWindowDaysAfter != nil {
		v := int(*req.BirthdayWindowDaysAfter)
		input.BirthdayWindowDaysAfter = &v
	}
	if req.ProxyHost != nil {
		input.ProxyHost = req.ProxyHost
	}
	if req.ProxyPort != nil {
		v := int(*req.ProxyPort)
		input.ProxyPort = &v
	}
	if req.ProxyUsername != nil {
		input.ProxyUsername = req.ProxyUsername
	}
	if req.ProxyPassword != nil {
		input.ProxyPassword = req.ProxyPassword
	}
	if req.ServerAliveExpectedCompletionTime != nil {
		v := int(*req.ServerAliveExpectedCompletionTime)
		input.ServerAliveExpectedCompletionTime = &v
	}
	if req.ServerAliveThresholdMinutes != nil {
		v := int(*req.ServerAliveThresholdMinutes)
		input.ServerAliveThresholdMinutes = &v
	}

	settings, err := h.svc.Update(r.Context(), tenantID, input)
	if err != nil {
		switch {
		case errors.Is(err, service.ErrInvalidBirthdayWindow):
			respondError(w, http.StatusBadRequest, err.Error())
		case errors.Is(err, service.ErrInvalidServerAliveTime):
			respondError(w, http.StatusBadRequest, err.Error())
		case errors.Is(err, service.ErrInvalidServerAliveThreshold):
			respondError(w, http.StatusBadRequest, err.Error())
		default:
			respondError(w, http.StatusInternalServerError, "Failed to update system settings")
		}
		return
	}

	// Audit log
	if h.auditService != nil {
		h.auditService.Log(r.Context(), r, service.LogEntry{
			TenantID:   tenantID,
			Action:     model.AuditActionUpdate,
			EntityType: "system_settings",
			EntityID:   settings.ID,
		})
	}

	respondJSON(w, http.StatusOK, mapSystemSettingsToResponse(settings))
}

// CleanupDeleteBookings handles POST /system-settings/cleanup/delete-bookings
func (h *SystemSettingsHandler) CleanupDeleteBookings(w http.ResponseWriter, r *http.Request) {
	tenantID, ok := middleware.TenantFromContext(r.Context())
	if !ok {
		respondError(w, http.StatusUnauthorized, "Tenant required")
		return
	}

	var req models.CleanupDeleteBookingsRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		respondError(w, http.StatusBadRequest, "Invalid request body")
		return
	}
	if err := req.Validate(nil); err != nil {
		respondError(w, http.StatusBadRequest, err.Error())
		return
	}

	input := service.CleanupDateRangeInput{
		DateFrom:    time.Time(*req.DateFrom),
		DateTo:      time.Time(*req.DateTo),
		EmployeeIDs: parseUUIDs(req.EmployeeIds),
		Confirm:     req.Confirm,
	}

	result, err := h.svc.DeleteBookings(r.Context(), tenantID, input)
	if err != nil {
		handleCleanupError(w, err)
		return
	}

	// Audit log for confirmed operations
	if !result.Preview && h.auditService != nil {
		h.auditService.Log(r.Context(), r, service.LogEntry{
			TenantID:   tenantID,
			Action:     model.AuditActionCleanup,
			EntityType: "bookings",
			EntityID:   tenantID, // tenant-level operation
		})
	}

	respondJSON(w, http.StatusOK, mapCleanupResult(result))
}

// CleanupDeleteBookingData handles POST /system-settings/cleanup/delete-booking-data
func (h *SystemSettingsHandler) CleanupDeleteBookingData(w http.ResponseWriter, r *http.Request) {
	tenantID, ok := middleware.TenantFromContext(r.Context())
	if !ok {
		respondError(w, http.StatusUnauthorized, "Tenant required")
		return
	}

	var req models.CleanupDeleteBookingDataRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		respondError(w, http.StatusBadRequest, "Invalid request body")
		return
	}
	if err := req.Validate(nil); err != nil {
		respondError(w, http.StatusBadRequest, err.Error())
		return
	}

	input := service.CleanupDateRangeInput{
		DateFrom:    time.Time(*req.DateFrom),
		DateTo:      time.Time(*req.DateTo),
		EmployeeIDs: parseUUIDs(req.EmployeeIds),
		Confirm:     req.Confirm,
	}

	result, err := h.svc.DeleteBookingData(r.Context(), tenantID, input)
	if err != nil {
		handleCleanupError(w, err)
		return
	}

	if !result.Preview && h.auditService != nil {
		h.auditService.Log(r.Context(), r, service.LogEntry{
			TenantID:   tenantID,
			Action:     model.AuditActionCleanup,
			EntityType: "booking_data",
			EntityID:   tenantID,
		})
	}

	respondJSON(w, http.StatusOK, mapCleanupResult(result))
}

// CleanupReReadBookings handles POST /system-settings/cleanup/re-read-bookings
func (h *SystemSettingsHandler) CleanupReReadBookings(w http.ResponseWriter, r *http.Request) {
	tenantID, ok := middleware.TenantFromContext(r.Context())
	if !ok {
		respondError(w, http.StatusUnauthorized, "Tenant required")
		return
	}

	var req models.CleanupReReadBookingsRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		respondError(w, http.StatusBadRequest, "Invalid request body")
		return
	}
	if err := req.Validate(nil); err != nil {
		respondError(w, http.StatusBadRequest, err.Error())
		return
	}

	input := service.CleanupDateRangeInput{
		DateFrom:    time.Time(*req.DateFrom),
		DateTo:      time.Time(*req.DateTo),
		EmployeeIDs: parseUUIDs(req.EmployeeIds),
		Confirm:     req.Confirm,
	}

	result, err := h.svc.ReReadBookings(r.Context(), tenantID, input)
	if err != nil {
		handleCleanupError(w, err)
		return
	}

	if !result.Preview && h.auditService != nil {
		h.auditService.Log(r.Context(), r, service.LogEntry{
			TenantID:   tenantID,
			Action:     model.AuditActionCleanup,
			EntityType: "bookings_reread",
			EntityID:   tenantID,
		})
	}

	respondJSON(w, http.StatusOK, mapCleanupResult(result))
}

// CleanupMarkDeleteOrders handles POST /system-settings/cleanup/mark-delete-orders
func (h *SystemSettingsHandler) CleanupMarkDeleteOrders(w http.ResponseWriter, r *http.Request) {
	tenantID, ok := middleware.TenantFromContext(r.Context())
	if !ok {
		respondError(w, http.StatusUnauthorized, "Tenant required")
		return
	}

	var req models.CleanupMarkDeleteOrdersRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		respondError(w, http.StatusBadRequest, "Invalid request body")
		return
	}
	if err := req.Validate(nil); err != nil {
		respondError(w, http.StatusBadRequest, err.Error())
		return
	}

	input := service.CleanupOrdersInput{
		OrderIDs: parseUUIDs(req.OrderIds),
		Confirm:  req.Confirm,
	}

	result, err := h.svc.MarkDeleteOrders(r.Context(), tenantID, input)
	if err != nil {
		handleCleanupError(w, err)
		return
	}

	if !result.Preview && h.auditService != nil {
		h.auditService.Log(r.Context(), r, service.LogEntry{
			TenantID:   tenantID,
			Action:     model.AuditActionCleanup,
			EntityType: "orders",
			EntityID:   tenantID,
		})
	}

	respondJSON(w, http.StatusOK, mapCleanupResult(result))
}

// --- Helpers ---

func handleCleanupError(w http.ResponseWriter, err error) {
	switch {
	case errors.Is(err, service.ErrInvalidDateRange):
		respondError(w, http.StatusBadRequest, err.Error())
	case errors.Is(err, service.ErrDateRangeTooLarge):
		respondError(w, http.StatusBadRequest, err.Error())
	case errors.Is(err, service.ErrCleanupNoOrderIDs):
		respondError(w, http.StatusBadRequest, err.Error())
	default:
		respondError(w, http.StatusInternalServerError, "Cleanup operation failed")
	}
}

func parseUUIDs(ids []strfmt.UUID) []uuid.UUID {
	result := make([]uuid.UUID, 0, len(ids))
	for _, id := range ids {
		parsed, err := uuid.Parse(id.String())
		if err == nil {
			result = append(result, parsed)
		}
	}
	return result
}

func mapSystemSettingsToResponse(s *model.SystemSettings) *models.SystemSettings {
	id := strfmt.UUID(s.ID.String())
	tenantID := strfmt.UUID(s.TenantID.String())
	createdAt := strfmt.DateTime(s.CreatedAt)
	updatedAt := strfmt.DateTime(s.UpdatedAt)

	resp := &models.SystemSettings{
		ID:                       &id,
		TenantID:                 &tenantID,
		RoundingRelativeToPlan:   s.RoundingRelativeToPlan,
		ErrorListEnabled:         s.ErrorListEnabled,
		TrackedErrorCodes:        s.TrackedErrorCodes,
		AutoFillOrderEndBookings: s.AutoFillOrderEndBookings,
		FollowUpEntriesEnabled:   s.FollowUpEntriesEnabled,
		ProxyEnabled:             s.ProxyEnabled,
		ServerAliveEnabled:       s.ServerAliveEnabled,
		ServerAliveNotifyAdmins:  s.ServerAliveNotifyAdmins,
		CreatedAt:                createdAt,
		UpdatedAt:                updatedAt,
	}

	before := int64(s.BirthdayWindowDaysBefore)
	resp.BirthdayWindowDaysBefore = &before
	after := int64(s.BirthdayWindowDaysAfter)
	resp.BirthdayWindowDaysAfter = &after

	if s.ProxyHost != nil {
		resp.ProxyHost = s.ProxyHost
	}
	if s.ProxyPort != nil {
		port := int64(*s.ProxyPort)
		resp.ProxyPort = &port
	}
	if s.ProxyUsername != nil {
		resp.ProxyUsername = s.ProxyUsername
	}
	if s.ServerAliveExpectedCompletionTime != nil {
		v := int64(*s.ServerAliveExpectedCompletionTime)
		resp.ServerAliveExpectedCompletionTime = &v
	}
	if s.ServerAliveThresholdMinutes != nil {
		v := int64(*s.ServerAliveThresholdMinutes)
		resp.ServerAliveThresholdMinutes = &v
	}

	return resp
}

func mapCleanupResult(r *service.CleanupResult) *models.CleanupResult {
	op := r.Operation
	count := r.AffectedCount
	now := strfmt.DateTime(time.Now())

	resp := &models.CleanupResult{
		Operation:     &op,
		AffectedCount: &count,
		Preview:       r.Preview,
		PerformedAt:   now,
	}

	if r.Details != nil {
		resp.Details = r.Details
	}

	return resp
}

File: src/trpc/routers/systemSettings.ts
/**
 * System Settings Router
 *
 * Provides singleton system settings get/update and cleanup operations via tRPC.
 *
 * Replaces the Go backend system settings endpoints:
 * - GET /system-settings -> systemSettings.get
 * - PUT /system-settings -> systemSettings.update
 * - POST /system-settings/cleanup/delete-bookings -> systemSettings.cleanupDeleteBookings
 * - POST /system-settings/cleanup/delete-booking-data -> systemSettings.cleanupDeleteBookingData
 * - POST /system-settings/cleanup/re-read-bookings -> systemSettings.cleanupReReadBookings
 * - POST /system-settings/cleanup/mark-delete-orders -> systemSettings.cleanupMarkDeleteOrders
 *
 * @see apps/api/internal/service/systemsettings.go
 */
import { z } from "zod"
import { createTRPCRouter, tenantProcedure } from "@/trpc/init"
import { requirePermission } from "@/lib/auth/middleware"
import { permissionIdByKey } from "@/lib/auth/permission-catalog"
import { handleServiceError } from "@/trpc/errors"
import * as settingsService from "@/lib/services/system-settings-service"

// --- Permission Constants ---

const SETTINGS_MANAGE = permissionIdByKey("settings.manage")!

// --- Output Schemas ---

const systemSettingsOutputSchema = z.object({
  id: z.string(),
  tenantId: z.string(),
  roundingRelativeToPlan: z.boolean(),
  errorListEnabled: z.boolean(),
  trackedErrorCodes: z.array(z.string()),
  autoFillOrderEndBookings: z.boolean(),
  birthdayWindowDaysBefore: z.number(),
  birthdayWindowDaysAfter: z.number(),
  followUpEntriesEnabled: z.boolean(),
  proxyHost: z.string().nullable(),
  proxyPort: z.number().nullable(),
  proxyUsername: z.string().nullable(),
  proxyEnabled: z.boolean(),
  serverAliveEnabled: z.boolean(),
  serverAliveExpectedCompletionTime: z.number().nullable(),
  serverAliveThresholdMinutes: z.number().nullable(),
  serverAliveNotifyAdmins: z.boolean(),
  createdAt: z.date(),
  updatedAt: z.date(),
})

const cleanupResultSchema = z.object({
  operation: z.string(),
  affectedCount: z.number(),
  preview: z.boolean(),
  details: z.record(z.string(), z.unknown()).optional(),
})

// --- Input Schemas ---

const updateSettingsInputSchema = z.object({
  roundingRelativeToPlan: z.boolean().optional(),
  errorListEnabled: z.boolean().optional(),
  trackedErrorCodes: z.array(z.string()).optional(),
  autoFillOrderEndBookings: z.boolean().optional(),
  birthdayWindowDaysBefore: z.number().int().min(0).max(90).optional(),
  birthdayWindowDaysAfter: z.number().int().min(0).max(90).optional(),
  followUpEntriesEnabled: z.boolean().optional(),
  proxyHost: z.string().max(255).nullable().optional(),
  proxyPort: z.number().int().nullable().optional(),
  proxyUsername: z.string().max(255).nullable().optional(),
  proxyPassword: z.string().max(255).nullable().optional(),
  proxyEnabled: z.boolean().optional(),
  serverAliveEnabled: z.boolean().optional(),
  serverAliveExpectedCompletionTime: z
    .number()
    .int()
    .min(0)
    .max(1439)
    .nullable()
    .optional(),
  serverAliveThresholdMinutes: z
    .number()
    .int()
    .min(1)
    .nullable()
    .optional(),
  serverAliveNotifyAdmins: z.boolean().optional(),
})

const cleanupDateRangeInputSchema = z.object({
  dateFrom: z.string().date(),
  dateTo: z.string().date(),
  employeeIds: z.array(z.string()).optional(),
  confirm: z.boolean(),
})

const cleanupOrdersInputSchema = z.object({
  orderIds: z.array(z.string()).min(1, "At least one order ID required"),
  confirm: z.boolean(),
})

// --- Helpers ---

/**
 * Maps a SystemSetting record to the output shape, omitting proxyPassword.
 */
function mapToOutput(s: Record<string, unknown>) {
  return {
    id: s.id as string,
    tenantId: s.tenantId as string,
    roundingRelativeToPlan: s.roundingRelativeToPlan as boolean,
    errorListEnabled: s.errorListEnabled as boolean,
    trackedErrorCodes: s.trackedErrorCodes as string[],
    autoFillOrderEndBookings: s.autoFillOrderEndBookings as boolean,
    birthdayWindowDaysBefore: s.birthdayWindowDaysBefore as number,
    birthdayWindowDaysAfter: s.birthdayWindowDaysAfter as number,
    followUpEntriesEnabled: s.followUpEntriesEnabled as boolean,
    proxyHost: (s.proxyHost as string | null) ?? null,
    proxyPort: (s.proxyPort as number | null) ?? null,
    proxyUsername: (s.proxyUsername as string | null) ?? null,
    // proxyPassword intentionally omitted (matches Go json:"-" behavior)
    proxyEnabled: s.proxyEnabled as boolean,
    serverAliveEnabled: s.serverAliveEnabled as boolean,
    serverAliveExpectedCompletionTime:
      (s.serverAliveExpectedCompletionTime as number | null) ?? null,
    serverAliveThresholdMinutes:
      (s.serverAliveThresholdMinutes as number | null) ?? null,
    serverAliveNotifyAdmins: s.serverAliveNotifyAdmins as boolean,
    createdAt: s.createdAt as Date,
    updatedAt: s.updatedAt as Date,
  }
}

// --- Router ---

export const systemSettingsRouter = createTRPCRouter({
  /**
   * systemSettings.get -- Returns the singleton system settings for the tenant.
   *
   * Creates default settings on first access (getOrCreate pattern).
   * ProxyPassword is never returned.
   *
   * Requires: settings.manage permission
   */
  get: tenantProcedure
    .use(requirePermission(SETTINGS_MANAGE))
    .output(systemSettingsOutputSchema)
    .query(async ({ ctx }) => {
      try {
        const settings = await settingsService.get(ctx.prisma, ctx.tenantId!)
        return mapToOutput(settings as unknown as Record<string, unknown>)
      } catch (err) {
        handleServiceError(err)
      }
    }),

  /**
   * systemSettings.update -- Partially updates system settings.
   *
   * Uses getOrCreate to ensure the singleton exists before updating.
   * ProxyPassword is never returned.
   *
   * Requires: settings.manage permission
   */
  update: tenantProcedure
    .use(requirePermission(SETTINGS_MANAGE))
    .input(updateSettingsInputSchema)
    .output(systemSettingsOutputSchema)
    .mutation(async ({ ctx, input }) => {
      try {
        const updated = await settingsService.update(
          ctx.prisma,
          ctx.tenantId!,
          input
        )
        return mapToOutput(updated as unknown as Record<string, unknown>)
      } catch (err) {
        handleServiceError(err)
      }
    }),

  /**
   * systemSettings.cleanupDeleteBookings -- Delete bookings in a date range.
   *
   * Preview mode (confirm: false): returns count of affected bookings.
   * Execute mode (confirm: true): deletes bookings and returns count.
   *
   * Requires: settings.manage permission
   */
  cleanupDeleteBookings: tenantProcedure
    .use(requirePermission(SETTINGS_MANAGE))
    .input(cleanupDateRangeInputSchema)
    .output(cleanupResultSchema)
    .mutation(async ({ ctx, input }) => {
      try {
        return await settingsService.cleanupDeleteBookings(
          ctx.prisma,
          ctx.tenantId!,
          input
        )
      } catch (err) {
        handleServiceError(err)
      }
    }),

  /**
   * systemSettings.cleanupDeleteBookingData -- Delete bookings, daily values,
   * and employee day plans in a date range.
   *
   * Preview mode: returns counts of bookings and daily values.
   * Execute mode: deletes all three entity types.
   *
   * Requires: settings.manage permission
   */
  cleanupDeleteBookingData: tenantProcedure
    .use(requirePermission(SETTINGS_MANAGE))
    .input(cleanupDateRangeInputSchema)
    .output(cleanupResultSchema)
    .mutation(async ({ ctx, input }) => {
      try {
        return await settingsService.cleanupDeleteBookingData(
          ctx.prisma,
          ctx.tenantId!,
          input
        )
      } catch (err) {
        handleServiceError(err)
      }
    }),

  /**
   * systemSettings.cleanupReReadBookings -- Re-read bookings in a date range.
   *
   * Preview mode: returns count of bookings.
   * Execute mode: recalculates bookings via RecalcService.
   *
   * Requires: settings.manage permission
   */
  cleanupReReadBookings: tenantProcedure
    .use(requirePermission(SETTINGS_MANAGE))
    .input(cleanupDateRangeInputSchema)
    .output(cleanupResultSchema)
    .mutation(async ({ ctx, input }) => {
      try {
        return await settingsService.cleanupReReadBookings(
          ctx.prisma,
          ctx.tenantId!,
          input
        )
      } catch (err) {
        handleServiceError(err)
      }
    }),

  /**
   * systemSettings.cleanupMarkDeleteOrders -- Delete orders by IDs.
   *
   * Preview mode: returns count of matching orders.
   * Execute mode: deletes the orders.
   * Uses Prisma model since Order has a Prisma model.
   *
   * Requires: settings.manage permission
   */
  cleanupMarkDeleteOrders: tenantProcedure
    .use(requirePermission(SETTINGS_MANAGE))
    .input(cleanupOrdersInputSchema)
    .output(cleanupResultSchema)
    .mutation(async ({ ctx, input }) => {
      try {
        return await settingsService.cleanupMarkDeleteOrders(
          ctx.prisma,
          ctx.tenantId!,
          input
        )
      } catch (err) {
        handleServiceError(err)
      }
    }),
})
